src/OpenTK.Windowing.GraphicsLibraryFramework/JoystickState.cs

```diff
@@ -0,0 +1,235 @@
+
+            _hats = new Hat[MaxHistory][];
+            _axes = new float[MaxHistory][];
+            _buttons = new byte[MaxHistory][];
+            for (int i = 0; i < MaxHistory; i++)
+            {
+                Array.Copy(source._hats[i], _hats[i], source._hats[i].Length);
+                Array.Copy(source._axes[i], _axes[i], source._axes[i].Length);
+                Array.Copy(source._buttons[i], _buttons[i], source._buttons[i].Length);
+            }
+
+            _current = source._current;
+            _previous = source._previous;
+        }
+
+        /// <inheritdoc />
+        public Hat GetHat(int index)
+        {
+            return _hats[_current][index];
+        }
+
+        /// <inheritdoc />
+        public Hat GetHatPrevious(int index)
+        {
+            return _hats[_previous][index];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SetHat(int index, Hat value)
+        {
+            _hats[_current][index] = value;
+        }
+
+        /// <inheritdoc />
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool IsButtonDown(int index)
+        {
+            byte b = _buttons[_current][index / 8];
+            int pow = (int)Math.Pow(2, index % 8);
+
+            return (b & pow) == pow;
+        }
+
+        /// <inheritdoc />
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool WasButtonDown(int index)
+        {
+            byte b0 = _buttons[_current][index / 8];
+            byte b1 = _buttons[_previous][index / 8];
+            int pow = (int)Math.Pow(2, index % 8);
+
+            return (b0 & pow) == pow && (b1 & pow) != pow;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SetButtonDown(int index, bool value)
+        {
+            int byteOffSet = index / 8;
+            int bitOffset = index % 8;
+
+            if (value)
+            {
+                _buttons[_current][byteOffSet] |= (byte)(1 << bitOffset);
+            }
+            else
+            {
+                _buttons[_current][byteOffSet] &= (byte)~(1 << bitOffset);
+            }
+        }
+
+        /// <inheritdoc />
+        public float GetAxis(int index)
+        {
+            return _axes[_current][index];
+        }
+
+        /// <inheritdoc />
+        public float GetAxisPrevious(int index)
+        {
+            return _axes[_previous][index];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SetAxis(int index, float value)
+        {
+            _axes[_current][index] = value < -1 ? -1 : (value > 1 ? 1 : value);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SetAxes(float[] axes)
+        {
+            _axes[_current] = axes;
+        }
+
+        /// <inheritdoc />
+        public override int GetHashCode()
+        {
+            int hashCode = _axes.GetHashCode();
```

Comment:
This should use the new `Hash.Combine()` API (or probably not implement this function at all)

Reply:
I'd prefer ditching any old code that we agree is not fit for purpose.